### PR TITLE
Fix a mem leak in evp_pkey_copy_downgraded()

### PR DESCRIPTION
**Pull request by mattcaswell**
If we get a failure during evp_pkey_copy_downgraded() and on entry *dest
was NULL then we leak the EVP_PKEY that was automatically allocated and
stored in *dest.

Found due to this comment:
https://github.com/openssl/openssl/pull/18355#issuecomment-1145028315

### CODE REVIEW

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
Ping for second review

**Comment by mattcaswell** [comment]
@openssl/committers

File: crypto/evp/p_lib.c
Comment (DDvO) [inline crypto/evp/p_lib.c:2057]: This guard is not really needed.
I suggest removing it, also for defensiveness such that `*dest = NULL` on error.

Comment (DDvO) [reply crypto/evp/p_lib.c:2057]: Oops, I had not taken into account that `*dest` may be non-NULL already initially.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**
